### ISSUE TITLE
Switch apis per request?

### ISSUE THREAD

**Issue by bitops** [remiprev/her#309]
With Her, is it possible to switch which Her::API object a model is using dynamically at runtime?

**Comment by giedriusr** [comment]
@bitops I am not sure if I correctly understood your question. How about others? Would you like to rewrite your question?

**Comment by bitops** [comment]
@giedriusr sorry, yes, it's not very clear.

I am working on a Rails application and all my models are Her models. The problem now is that this system is going to be multi-tenant, and each tenant has their own API instance. E.g. for Tenant A we talk to ```https://some-api-host/``` and for Tenant B we talk to ```https://some-other-api-host/```.

So my need is to be able to change the url dynamically based on which tenant I'm fetching data for. Is that more clear?

**Comment by giedriusr** [comment]
Tenant in this case is a record from the database? Because Her supports multi API per model, you can find it in the documentation. Multi API per DB record, I doubt it. Try to poke maintainers of this gem.

**Comment by bitops** [comment]
@giedriusr Tenant is the user of the system. Said differently:

```ruby
Foo.get("/important_data") # => needs to talk to https://some-api-host/
Foo.get("/important_data") # => needs to talk to https://some-other-api-host/
```

**Comment by bitops** [comment]
@giedriusr looking at the docs, it's supporting a different use case than what I have. I need to point to a different API for the same class/model.

**Comment by bitops** [comment]
@remiprev thoughts?

**Comment by tomash** [comment]
I'd also be interested in this (similar problem, I'm using it for "meta-admin" application that talks to endpoints of all the tenants' apps)

**Comment by abhikalakuntla** [comment]
This was an experimental solution we created - I'm not too sure this is the best way though.

In our use case, we had to switch out the endpoint based on the user that was logged in. User-type A would hit endpoint of "a-user.com/api" and user-type B would hit "b-user.com/api". The user type was able to be extracted based on user auth token. 

So in our Her config, we setup a API endpoint configuration for each type of user. So that we can say `Model.use_api (User-A-API)` or `Model.use_api(User-B-API)`. In that format you can also customize the middlewares you want for each api endpoint. 

Then in our controllers, we have a before_filter that will change the direction of the model based on the user auth token. 

But you might get race conditions with this solution. The issue is: let's say request 1 (user type a) hits it's api but the request takes 20sec. Then you have request 2 (user type b) hit it's api a little after request 1. If both request 1 and request 2 use the same model, then you might run into issues of timing seeing as to how models are global. However, our house was running this setup under unicorn, which makes a copy of all models for each worker AND, in general, ruby processes it's requests serially. 

If you want to be super sure and want to correct this (DISCLAIMER: have not done this myself), you can generate a copy of each model per api endpoint and hash it in a global space (yeah, saving to a global accessible space is generally frowned upon but maybe okay because a model is not supposed to be dynamic). Then when a request comes in you assign the local model variable to the desired one you stored in the hash.

**Comment by hubert** [comment]
i'm not clear on the use case based on the OP question, but how do you determine which API you want a given object to query. what is the reason that you cannot have two models one for talking to each of the APIs and then use the same logic that determines which API to use to select which Her::Model to use. 

if there are many APIs to hit, and these can only be determined at runtime, you could create a Factory class that builds a Her::Model class (perhaps based on a class that contains any common logic) that is configured to speak to whichever API your business logic dictates it should. 

let me know if i am totally misunderstanding what is desired.

**Comment by wmulligan87** [comment]
Hi, I just wanted to resurrect this issue as I'm looking to use the Her gem for a customer portal app I'm building, but I have a scenario where each request, depending on the current_user logged in user, will be accessing a different url to fetch the json data.

The app is multi-tenanted, where there are a bunch of Users which belong to Companies (Company has_many Users), and those companies have their own API for fetching data to show the user.

For example user X belongs to company 1, user Y belongs to company 2.
User log in, and want to see an index of (for example...) 'Products' from '<user.company.url>/api/products.json'.

So, User X logs in and gets an index of products, fetched from 'http://company1.com/api/products.json'
User Y then goes to the same index, but the products are fetched from 'http://company2.com/api/products.json'

The app has to handle a large quantity of companies, these aren't constants, so I can't hardcode different API's in the initializers as mentioned in the docs.

So basically, I need to be able to change the URL per-request. I can't see a documented solution for something like this in Her. I was originally looking at ActiveResource - which can change site's per request via modifying a class variable per request (obviously not ideal, but it worked in testing) - but it hasn't been maintained for a long time so seriously want to avoid that...

Any thoughts? Could I achieve this with Her?

**Comment by bitops** [comment]
Just following up here since it seems like folks are still asking this question. I think if this is a problem that affects you as a developer, you should not be using Her in your application. 

Her is really great if you are trying to get something done in a quick and dirty way and/or if you have a large number of integrations for a mash-up type application.

Her is not a good fit if you're building out an enterprise backend where a strong and consistent domain model is required. Particularly if the responses you get on different request types have very different shapes. In my own case, the team I was on at the time ended up writing a proper SDK / gem for the API we were using and removing Her. It was a lot of work but we ended up with a well-understood domain model and it was much easier to test things. 

The thing that forced our hand in the switch is that the API we were coding against had really different response formats on success versus error. Trying to build a consistent domain model to accommodate this with Her was much more trouble than it was really worth.

None of this should be construed as a criticism of Her. It's a library that works well and it's a perfect fit for a certain set of use cases. But, if you find yourself on this thread _really wishing_ you had this functionality, you might consider finding a different solution.

**Comment by bitops** [comment]
I'm going to close out this issue because I don't think it should be a responsibility or design goal of Her to work in this way.
